### PR TITLE
[9.x] Add env processors (for use in config files)

### PR DESCRIPTION
**Pull request by X-Coder264**
One feature that I've really missed so far in the framework are env processors.

Consider this fairly basic example:

```php
// app.php
return [
    'amount_of_foo' => env('AMOUNT_OF_FOO', 5),
];

// somewhere in the codebase
private function getFooAmount(): int
{
    return config('app.amount_of_foo');
}
```

The issue here is that if the `AMOUNT_OF_FOO` env variable does not exist in your `.env` file everything works, but if it exists everything breaks because the value returned is a string so you get a `TypeError` (and I've been bitten by this plenty of times, especially when [binding primitives using the container](https://laravel.com/docs/8.x/container#binding-primitives)).

Of course this is fixable by doing this:

```diff
// app.php
return [
-  'amount_of_foo' => env('AMOUNT_OF_FOO', 5),
+  'amount_of_foo' => (int) env('AMOUNT_OF_FOO', 5),
];
```

But this is not very pretty and it's very cumbersome to remember having to do this every time when the expected return type is not a string.

This PR introduces env processors which allow you to do this:

```diff
// app.php
return [
-  'amount_of_foo' => env('AMOUNT_OF_FOO', 5), // same as Env::get('AMOUNT_OF_FOO', 5);
+  'amount_of_foo' => Env::integer('AMOUNT_OF_FOO', 5),
];
```

The code becomes more readable and we ensure type safety here which greatly improves DX.

**Env processors are far more powerful than this though - currently you cannot natively get anything but a string (or null) back from a `env()` call.**

Let's take the (real world) example from https://timacdonald.me/introducing-lazy-loading-prevention-into-an-existing-laravel-application/ and show a before and after with the usage of env processors:

```diff
return [
-    'lazy_loading_reporting_lottery' => array_map(
-         'intval',
-         explode(',', env('LOG_LAZY_LOADING_DETECTION_LOTTERY', '1,500'))
-     ),
+    'lazy_loading_reporting_lottery' => Env::getWithProcessor(
+            EnvProcessorFactory::arrayProcessor()->integerElements(),
+            'LOG_LAZY_LOADING_DETECTION_LOTTERY',
+            [1, 500],
        );
];
```

If we didn't need to have integer elements the above call could've been just
```php
Env::array('LOG_LAZY_LOADING_DETECTION_LOTTERY', ['1', '500']),
```

The delimiter is customizable too (and defaults to `,`).

## Added env processors

`Env::array()` -> ensures that an array of values is always returned, can be configured to cast all elements to integers or floats if necessary, the delimiter is customizable too
`Env::string()` -> ensures that a string is always returned
`Env::integer()` -> ensures that an integer is always returned
`Env::float()` -> ensures that a float is always returned
`Env::boolean()` -> ensures that a boolean is always returned
`Env::trim()` -> trims the value (useful in combination with the `file` processor)
`Env::file()` -> returns the string content of the file at the given path
`Env::json()` -> returns the JSON decoded content of the given string
`Env::base64()` -> returns the base64 URL decoded content of the given string
`Env::chain()` -> runs the env value through two or more of the above processors before returning it

## Some usage examples

```php
// some_config_file.php

return [
    'foo' => Env::array('FOO'), // returns an array
    'bar' => Env::boolean('BAR'), // returns a boolean
    'qux' => Env::json('QUX'), // returns the decoded content (array or null)
    'quux' => Env::chain('QUUX', null, EnvProcessorFactory::fileProcessor(), EnvProcessorFactory::jsonProcessor()), // gets the file content from the file path, JSON decodes that string and returns it
    'quuz' => Env::getWithProcessor(EnvProcessorFactory::arrayProcessor()->delimiter(';')->floatElements(), 'QUUZ'), // returns an array
];
```

In order to prevent bugs all processors are immutable objects (this means that if a processor is created in a config file it can safely be reused in that same file), for example:

```php
// some_config_file.php

// FOO=1 (or FOO=true)

$booleanProcessor = EnvProcessorFactory::booleanProcessor();

return [
    'x' => Env::getWithProcessor($booleanProcessor->not(), 'FOO'), // calling not() here tells the processor to negate the value before returning it
    'y' => Env::getWithProcessor($booleanProcessor, 'FOO'),
];
```

In the above example `x` is `false` while `y` is `true`. If the processors weren't immutable both `x` and `y` would've been `false`.

Just to clarify, `Env::boolean('FOO')` is syntax sugar for `Env::getWithProcessor(EnvProcessorFactory::booleanProcessor(), 'FOO', true)`. The only reason why you couldn't use this syntax sugar is when you need to use the more advanced features that a processor exposes (like negating the value in case of the boolean processor, or like adjusting the delimiter for the array processor etc.)

## Conclusion

Env var processors is a powerful feature (which brings type safety and reduces a lot of code which is usually copy pasted all the time in config files) which IMO greatly improves DX and as such should be a part of the framework. The inspiration for this PR are the equivalent env var processors from Symfony -> https://symfony.com/doc/current/configuration/env_var_processors.html

There are no breaking changes here and the code is fully typed to ensure the aforementioned type safety.

User apps and Laravel packages can create their own processors by implementing the `\Illuminate\Support\EnvProcessors\EnvProcessorInterface` contract.

### CODE REVIEW

File: src/Illuminate/Support/Env.php [hunk 1]
```diff
@@ -1,9 +1,12 @@
 <?php
 
+declare(strict_types=1);
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/Env.php:3]: don't add this

File: src/Illuminate/Support/EnvProcessors/BooleanProcessor.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
+<?php
+
+declare(strict_types=1);
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/EnvProcessors/BooleanProcessor.php:3]: same for all files

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Support/Env.php (hunk 1 above)
Comment (X-Coder264) [reply src/Illuminate/Support/Env.php:3]: removed

File: src/Illuminate/Support/EnvProcessors/BooleanProcessor.php (hunk 2 above)
Comment (X-Coder264) [reply src/Illuminate/Support/EnvProcessors/BooleanProcessor.php:3]: removed

**Comment by driesvints** [comment]
While I definitely appreciate the thought behind this PR (and of course all the other work you've done so far for Laravel) I think this PR unnecessarily overcomplicates things. You mention at some point that things are fixable with a typecast but then say:

> But this is not very pretty and it's very cumbersome to remember having to do this every time when the expected return type is not a string.

I don't agree with this and think we should just rely on the PHP type system instead of creating an abstract layer on top of it. This is a whole lot of extra code for us to maintain while we could just expect/require people to use the proper types/values.

I can of course be that I'm not seeing the main advantages of this. Your link to Symfony's implementation is a good indication but I still believe this is a bit too much, sorry.

**Comment by X-Coder264** [comment]
Not having to think about the type is just one of the benefits of this. Currently `env()` returns only strings and I think that's quite a limitation so what this brings to the table is stuff like conversion to arrays (and stuff like chaining the processors). Yes, this doesn't bring nothing new to the table that can't be already done, but since Laravel is all about DX (and other frameworks have this functionality too) I thought it was worth to try to PR this.

I know that I've used this feature in Symfony apps a bunch of times and that I've missed this a lot of times in my Laravel projects too (otherwise I wouldn't have bothered to PR this).

Regarding the maintenance burden, I think it's close to zero. Yes, there's a "lot" of classes, but it's a couple of lines of logic per class at most. The Symfony env var processors haven't changed much, if at all since they were introduced (there were maybe just a couple of new processors added). The entire code for `env` hasn't been touched in any significant way for a long period of time and I don't expect this to change any more than that really (as there's only so much ways you can convert a string into a boolean or an array etc.).

**Comment by taylorotwell** [comment]
I honestly find adding a simple, built-in PHP cast to your statement much easier than adding an entirely new feature to support env processors.

```
$foo = (int) $bar;
```

**Comment by morloderex** [comment]
@X-Coder264 Would it be possible to release this as a package sense it has been rejected?

**Comment by X-Coder264** [comment]
@morloderex Sure, I can do that. I can't promise anything until Sunday or probably some time next week though.

**Pull request closed without merging**